File: client/web-client/src/jsMain/kotlin/ro/jf/funds/client/web/api/AccountApi.kt
package ro.jf.funds.client.web.api

import com.benasher44.uuid.uuidFrom
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.promise
import ro.jf.funds.client.sdk.AccountClient
import ro.jf.funds.client.web.model.JsAccount
import ro.jf.funds.fund.api.model.AccountName
import ro.jf.funds.fund.api.model.CreateAccountTO
import ro.jf.funds.platform.api.model.FinancialUnit
import kotlin.js.JsExport
import kotlin.js.Promise

@JsExport
object AccountApi {
    private val config = js("window.FUNDS_CONFIG")
    private val fundServiceUrl: String = config?.fundServiceUrl as? String ?: "http://localhost:5253"
    private val accountClient = AccountClient(baseUrl = fundServiceUrl)

    fun listAccounts(userId: String): Promise<Array<JsAccount>> = GlobalScope.promise {
        val uuid = uuidFrom(userId)
        val accounts = accountClient.listAccounts(uuid)
        accounts.map {
            JsAccount(
                id = it.id.toString(),
                name = it.name.toString(),
                unitType = it.unit.type.value,
                unitValue = it.unit.value,
            )
        }.toTypedArray()
    }

    fun createAccount(
        userId: String,
        name: String,
        unitType: String,
        unitValue: String,
    ): Promise<JsAccount> = GlobalScope.promise {
        val uuid = uuidFrom(userId)
        val request = CreateAccountTO(
            name = AccountName(name),
            unit = FinancialUnit.of(unitType, unitValue),
        )
        val account = accountClient.createAccount(uuid, request)
        JsAccount(
            id = account.id.toString(),
            name = account.name.toString(),
            unitType = account.unit.type.value,
            unitValue = account.unit.value,
        )
    }

    fun deleteAccount(userId: String, accountId: String): Promise<Unit> = GlobalScope.promise {
        accountClient.deleteAccount(uuidFrom(userId), uuidFrom(accountId))
    }
}

File: client/web-client-react/src/jsMain/resources/react/api/accountApi.ts
import { PageResponse, PaginationParams, SortParams, AccountSortField } from './types';

export interface Account {
    id: string;
    name: string;
    unit: {
        type: string;
        value: string;
    };
}

export interface ListAccountsParams {
    pagination?: PaginationParams;
    sort?: SortParams<AccountSortField>;
}

export interface ListAccountsResult {
    items: Account[];
    total: number;
}

interface CreateAccountRequest {
    name: string;
    unit: {
        type: string;
        value: string;
    };
}

declare const window: Window & {
    FUNDS_CONFIG?: { fundServiceUrl?: string };
};

function getBaseUrl(): string {
    const url = window.FUNDS_CONFIG?.fundServiceUrl;
    if (!url) {
        throw new Error('FUNDS_CONFIG.fundServiceUrl is not configured');
    }
    return url;
}

const BASE_PATH = '/funds-api/fund/v1';

export async function listAccounts(
    userId: string,
    params?: ListAccountsParams
): Promise<ListAccountsResult> {
    const queryParams = new URLSearchParams();

    if (params?.pagination) {
        queryParams.set('offset', params.pagination.offset.toString());
        queryParams.set('limit', params.pagination.limit.toString());
    }

    if (params?.sort) {
        queryParams.set('sort', params.sort.field);
        queryParams.set('order', params.sort.order);
    }

    const queryString = queryParams.toString();
    const url = `${getBaseUrl()}${BASE_PATH}/accounts${queryString ? `?${queryString}` : ''}`;

    const response = await fetch(url, {
        headers: { 'FUNDS_USER_ID': userId }
    });
    if (!response.ok) throw new Error(`Failed to list accounts: ${response.status}`);
    const data: PageResponse<Account> = await response.json();
    return { items: data.items, total: data.total };
}

export async function createAccount(
    userId: string,
    name: string,
    unitType: string,
    unitValue: string
): Promise<Account> {
    const response = await fetch(`${getBaseUrl()}${BASE_PATH}/accounts`, {
        method: 'POST',
        headers: {
            'FUNDS_USER_ID': userId,
            'Content-Type': 'application/json'
        },
        body: JSON.stringify({
            name,
            unit: { type: unitType, value: unitValue }
        } as CreateAccountRequest)
    });
    if (!response.ok) throw new Error(`Failed to create account: ${response.status}`);
    return response.json();
}

export async function deleteAccount(userId: string, accountId: string): Promise<void> {
    const response = await fetch(`${getBaseUrl()}${BASE_PATH}/accounts/${accountId}`, {
        method: 'DELETE',
        headers: { 'FUNDS_USER_ID': userId }
    });
    if (!response.ok) throw new Error(`Failed to delete account: ${response.status}`);
}
